common/flatpak-prune.c

```diff
@@ -0,0 +1,826 @@
+      }
+    }
+
+  return TRUE;
+}
+
+typedef struct {
+  OstreeRepo *repo;
+  FlatpakOstreeObjectNameBag *reachable;
+  gboolean dont_prune;
+  guint n_reachable;
+  guint n_unreachable;
+  guint64 freed_bytes;
+} OtPruneData;
+
+static gboolean
+prune_loose_object (OtPruneData          *data,
+                    const char           *checksum,
+                    OstreeObjectType      objtype,
+                    GCancellable         *cancellable,
+                    GError              **error)
+{
+  guint64 storage_size = 0;
+
+  flatpak_debug2 ("Pruning unneeded object %s.%s", checksum,
+                  ostree_object_type_to_string (objtype));
+
+  if (!ostree_repo_query_object_storage_size (data->repo, objtype, checksum,
+                                              &storage_size, cancellable, error))
+    return FALSE;
+
+  data->freed_bytes += storage_size;
+  data->n_unreachable++;
+
+  if (!data->dont_prune)
+    {
+      if (objtype == OSTREE_OBJECT_TYPE_COMMIT)
+        {
+          if (!remove_extra_commitmeta (data->repo, checksum, cancellable, error))
+            return FALSE;
+
+          if (!ostree_repo_mark_commit_partial (data->repo, checksum, FALSE, error))
+            return FALSE;
+        }
+
+      if (!ostree_repo_delete_object (data->repo, objtype, checksum,
+                                      cancellable, error))
+        return FALSE;
+    }
+
+  return TRUE;
+}
+
+static gboolean
+prune_unreachable_loose_objects_at (OstreeRepo             *self,
+                                    OtPruneData            *data,
+                                    int                     dfd,
+                                    const char             *prefix,
+                                    GCancellable           *cancellable,
+                                    GError                **error)
+{
+
+  g_auto(GLnxDirFdIterator) dfd_iter = { 0, };
+  gboolean exists;
+  if (!ot_dfd_iter_init_allow_noent (dfd, prefix, &dfd_iter, &exists, error))
+    return FALSE;
+  /* Note early return */
+  if (!exists)
+    return TRUE;
+
+  while (TRUE)
+    {
+      struct dirent *dent;
+      FlatpakOstreeObjectName key;
+
+      if (!glnx_dirfd_iterator_next_dent (&dfd_iter, &dent, cancellable, error))
+        return FALSE;
+      if (dent == NULL)
+        break;
+
+      const char *name = dent->d_name;
+      if (strcmp (name, ".") == 0 ||
+          strcmp (name, "..") == 0)
+        continue;
+
+      const char *dot = strrchr (name, '.');
+      if (!dot)
+        continue;
+
+      OstreeObjectType objtype;
+
+      if (strcmp (dot, ".filez") == 0)
+        objtype = OSTREE_OBJECT_TYPE_FILE;
+      else if (strcmp (dot, ".dirtree") == 0)
+        objtype = OSTREE_OBJECT_TYPE_DIR_TREE;
+      else if (strcmp (dot, ".dirmeta") == 0)
+        objtype = OSTREE_OBJECT_TYPE_DIR_META;
+      else if (strcmp (dot, ".commit") == 0)
+        objtype = OSTREE_OBJECT_TYPE_COMMIT;
+      else
```

Comment:
`OSTREE_OBJECT_TYPE_PAYLOAD_LINK` isn't here but not sure if we'll run into it

Reply:
That only happens in bare repos.